File: cs205/jpetroulesCONDO/jpetroulesCONDOcs/MainForm.cs
using System;
using System.Drawing;
using System.Globalization;
using System.Net.Mail;
using System.Text;
using System.Windows.Forms;

namespace jpetroulesCONDOcs
{
    public partial class MainForm : Form
    {
        public MainForm()
        {
            this.InitializeComponent();
        }

        /// <summary>
        /// Generates a receipt containing customer information, rental unit, rental period,
        /// optional services and payment information and returns the result as a string.
        /// </summary>
        private string GenerateReceipt()
        {
            StringBuilder builder = new StringBuilder();

            // Add the customer information
            builder.AppendLine(string.Format(CultureInfo.CurrentCulture, "{0} {1} {2}", this.firstNameTextBox.Text, this.middleNameTextBox.Text, this.lastNameTextBox.Text));
            builder.AppendLine(this.streetAddressTextBox.Text);
            builder.AppendLine(string.Format(CultureInfo.CurrentCulture, "{0}, {1} {2}", this.cityTextBox.Text, this.stateComboBox.Text, this.zipCodeMaskedTextBox.MaskedTextProvider.ToString()));
            builder.AppendLine(this.phoneMaskedTextBox.MaskedTextProvider.ToString());
            builder.AppendLine(this.emailAddressTextBox.Text);
            builder.AppendLine();

            // Add the rental unit information
            if (this.studioRadioButton.Checked)
            {
                builder.AppendLine(this.studioRadioButton.Text);
            }
            else if (this.suite2RadioButton.Checked)
            {
                builder.AppendLine(this.suite2RadioButton.Text);
            }
            else if (this.suite3RadioButton.Checked)
            {
                builder.AppendLine(this.suite3RadioButton.Text);
            }
            else if (this.penthouseRadioButton.Checked)
            {
                builder.AppendLine(this.penthouseRadioButton.Text);
            }

            // Add the rental period information (check-in/check-out times)
            TimeSpan rentalPeriod = this.checkOutDateTimePicker.Value.Date - this.checkInDateTimePicker.Value.Date;
            builder.AppendLine(string.Format(CultureInfo.CurrentCulture, "Check in: {0}", this.checkInDateTimePicker.Value.Date));
            builder.AppendLine(string.Format(CultureInfo.CurrentCulture, "Check out: {0}", this.checkOutDateTimePicker.Value.Date));
            builder.AppendLine(string.Format(CultureInfo.CurrentCulture, "Rental period: {0} days", rentalPeriod.Days));
            builder.AppendLine();

            // Add the optional services if they were selected
            if (this.maidServiceCheckBox.Checked || this.linenServiceCheckBox.Checked || this.limoServiceCheckBox.Checked)
            {
                builder.AppendLine("Optional services:");
            }

            if (this.maidServiceCheckBox.Checked)
            {
                builder.AppendLine(this.maidServiceCheckBox.Text);
            }

            if (this.linenServiceCheckBox.Checked)
            {
                builder.AppendLine(this.linenServiceCheckBox.Text);
            }

            if (this.limoServiceCheckBox.Checked)
            {
                builder.AppendLine(this.limoServiceCheckBox.Text);
            }

            // Add the customer's total bill
            builder.AppendLine(string.Format(CultureInfo.CurrentCulture, "Grand total: {0:C}", this.CalculateTotal()));

            return builder.ToString();
        }

        /// <summary>
        /// Calculates the customer's total bill and returns the result as a decimal.
        /// </summary>
        private decimal CalculateTotal()
        {
            TimeSpan rentalPeriod = this.checkOutDateTimePicker.Value.Date - this.checkInDateTimePicker.Value.Date;

            // The total cost of the customer's condo rental
            decimal totalCost = 0;

            // Determine the weekly price of the selected rental unit
            decimal weeklyUnitPrice = 0;
            if (this.studioRadioButton.Checked)
            {
                weeklyUnitPrice = 1000;
            }
            else if (this.suite2RadioButton.Checked)
            {
                weeklyUnitPrice = 2000;
            }
            else if (this.suite3RadioButton.Checked)
            {
                weeklyUnitPrice = 3000;
            }
            else if (this.penthouseRadioButton.Checked)
            {
                weeklyUnitPrice = 15000;
            }

            // Add the cost of the unit for the amount of days in the rental period
            totalCost += (weeklyUnitPrice / 7) * rentalPeriod.Days;

            // If they're renting for 3 months or more, give them a 20% discount
            if (rentalPeriod.Days >= 90)
            {
                totalCost *= 0.8m;
            }

            // Maid service is $100/day
            if (this.maidServiceCheckBox.Checked)
            {
                totalCost += 100 * rentalPeriod.Days;
            }

            // The penthouse has 4 bedrooms; the cost is $20/day per bedroom, so we multiply by 4 bedrooms
            // if the penthouse was selected, otherwise just multiply by 1 bedroom
            if (this.linenServiceCheckBox.Checked)
            {
                totalCost += 20 * rentalPeriod.Days * (this.penthouseRadioButton.Checked ? 4 : 1);
            }

            // Limo service is $75/day
            if (this.limoServiceCheckBox.Checked)
            {
                totalCost += 75 * rentalPeriod.Days;
            }

            // Vermont has a 5% rooms and meals tax
            return totalCost * 1.05m;
        }

        /// <summary>
        /// Determines whether the user has entered complete and valid input in all fields.
        /// </summary>
        private bool IsInputValid()
        {
            return
                this.ValidateForControl(this.firstNameTextBox) &
                this.ValidateForControl(this.middleNameTextBox) &
                this.ValidateForControl(this.lastNameTextBox) &
                this.ValidateForControl(this.streetAddressTextBox) &
                this.ValidateForControl(this.cityTextBox) &
                this.ValidateForControl(this.stateComboBox) &
                this.ValidateForControl(this.zipCodeMaskedTextBox) &
                this.ValidateForControl(this.phoneMaskedTextBox) &
                this.ValidateForControl(this.emailAddressTextBox) &
                this.ValidateForControl(this.checkInDateTimePicker) &
                this.ValidateForControl(this.checkOutDateTimePicker) &
                this.ValidateForControl(this.rentalUnitsGroupBox);
        }

        /// <summary>
        /// Runs validation for the specified control and returns whether its corresponding input is valid.
        /// </summary>
        /// <param name="control">The control to validate.</param>
        private bool ValidateForControl(Control control)
        {
            TextBox firstNameTextBox = control as TextBox;
            if (firstNameTextBox == this.firstNameTextBox)
            {
                if (string.IsNullOrWhiteSpace(this.firstNameTextBox.Text))
                {
                    this.firstNameErrorProvider.SetError(this.firstNameTextBox, "Please enter a first name.");
                    return false;
                }
                else
                {
                    this.firstNameErrorProvider.Clear();
                    return true;
                }
            }

            TextBox middleNameTextBox = control as TextBox;
            if (middleNameTextBox == this.middleNameTextBox)
            {
                if (string.IsNullOrWhiteSpace(this.middleNameTextBox.Text))
                {
                    this.middleNameErrorProvider.SetError(this.middleNameTextBox, "Please enter a middle name.");
                    return false;
                }
                else
                {
                    this.middleNameErrorProvider.Clear();
                    return true;
                }
            }

            TextBox lastNameTextBox = control as TextBox;
            if (lastNameTextBox == this.lastNameTextBox)
            {
                if (String.IsNullOrWhiteSpace(this.lastNameTextBox.Text))
                {
                    this.lastNameErrorProvider.SetError(this.lastNameTextBox, "Please enter a last name.");
                    return false;
                }
                else
                {
                    this.lastNameErrorProvider.Clear();
                    return true;
                }
            }

            TextBox streetAddressTextBox = control as TextBox;
            if (streetAddressTextBox == this.streetAddressTextBox)
            {
                if (String.IsNullOrWhiteSpace(this.streetAddressTextBox.Text))
                {
                    this.streetAddressErrorProvider.SetError(this.streetAddressTextBox, "Please enter a street address.");
                    return false;
                }
                else
                {
                    this.streetAddressErrorProvider.Clear();
                    return true;
                }
            }

            TextBox cityTextBox = control as TextBox;
            if (cityTextBox == this.cityTextBox)
            {
                if (String.IsNullOrWhiteSpace(this.cityTextBox.Text))
                {
                    this.cityErrorProvider.SetError(this.cityTextBox, "Please enter a city.");
                    return false;
                }
                else
                {
                    this.cityErrorProvider.Clear();
                    return true;
                }
            }

            ComboBox stateComboBox = control as ComboBox;
            if (stateComboBox == this.stateComboBox)
            {
                if (this.stateComboBox.SelectedIndex < 0)
                {
                    this.stateErrorProvider.SetError(this.stateComboBox, "Please select a state.");
                    return false;
                }
                else
                {
                    this.stateErrorProvider.Clear();
                    return true;
                }
            }

            MaskedTextBox zipCodeMaskedTextBox = control as MaskedTextBox;
            if (zipCodeMaskedTextBox == this.zipCodeMaskedTextBox)
            {
                if (this.zipCodeMaskedTextBox.MaskCompleted)
                {
                    this.zipCodeErrorProvider.Clear();
                    return true;
                }
                else
                {
                    this.zipCodeErrorProvider.SetError(this.zipCodeMaskedTextBox, "Please enter a valid ZIP code.");
                    return false;
                }
            }

            MaskedTextBox phoneMaskedTextBox = control as MaskedTextBox;
            if (phoneMaskedTextBox == this.phoneMaskedTextBox)
            {
                if (this.phoneMaskedTextBox.MaskCompleted)
                {
                    this.phoneErrorProvider.Clear();
                    return true;
                }
                else
                {
                    this.phoneErrorProvider.SetError(this.phoneMaskedTextBox, "Please enter a valid telephone number.");
                    return false;
                }
            }

            TextBox emailAddressTextBox = control as TextBox;
            if (emailAddressTextBox == this.emailAddressTextBox)
            {
                MailAddress email;
                try
                {
                    email = new MailAddress(this.emailAddressTextBox.Text);
                    this.emailAddressErrorProvider.Clear();
                    return true;
                }
                catch (ArgumentException)
                {
                    this.emailAddressErrorProvider.SetError(this.emailAddressTextBox, "Please enter an email address.");
                    return false;
                }
                catch (FormatException)
                {
                    this.emailAddressErrorProvider.SetError(this.emailAddressTextBox, "Please enter a valid email address.");
                    return false;
                }
            }

            DateTimePicker checkInDateTimePicker = control as DateTimePicker;
            DateTimePicker checkOutDateTimePicker = control as DateTimePicker;
            if (checkInDateTimePicker == this.checkInDateTimePicker || checkOutDateTimePicker == this.checkOutDateTimePicker)
            {
                TimeSpan rentalPeriod = this.checkOutDateTimePicker.Value.Date - this.checkInDateTimePicker.Value.Date;
                if (rentalPeriod.TotalDays > 0)
                {
                    this.rentalTermLabel.Text = string.Format(CultureInfo.CurrentCulture, "You have selected a rental period of {0} days", rentalPeriod.Days);
                    this.rentalTermLabel.ForeColor = Label.DefaultForeColor;
                    return true;
                }
                else
                {
                    this.rentalTermLabel.Text = "Check in time cannot be equal to or later than check out time";
                    this.rentalTermLabel.ForeColor = Color.Red;
                    return false;
                }
            }

            GroupBox rentalUnitsGroupBox = control as GroupBox;
            if (rentalUnitsGroupBox == this.rentalUnitsGroupBox)
            {
                bool oneButtonChecked = this.studioRadioButton.Checked ^ this.suite2RadioButton.Checked ^ this.suite3RadioButton.Checked ^ this.penthouseRadioButton.Checked;
                if (oneButtonChecked)
                {
                    this.rentalUnitsErrorProvider.Clear();
                    return true;
                }
                else
                {
                    this.rentalUnitsErrorProvider.SetError(this.rentalUnitsGroupBox, "Please select a rental unit.");
                    return false;
                }
            }

            // An invalid control can just return false
            return false;
        }

        private void generateReceiptButton_Click(object sender, EventArgs e)
        {
            if (this.IsInputValid())
            {
                this.outputRichTextBox.Text = this.GenerateReceipt();
            }
            else
            {
                MessageBox.Show("Please correct the input errors before generating a receipt.", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
        }

        private void firstNameTextBox_Validating(object sender, System.ComponentModel.CancelEventArgs e)
        {
            this.ValidateForControl(this.firstNameTextBox);
        }

        private void middleNameTextBox_Validating(object sender, System.ComponentModel.CancelEventArgs e)
        {
            this.ValidateForControl(this.middleNameTextBox);
        }

        private void lastNameTextBox_Validating(object sender, System.ComponentModel.CancelEventArgs e)
        {
            this.ValidateForControl(this.lastNameTextBox);
        }

        private void streetAddressTextBox_Validating(object sender, System.ComponentModel.CancelEventArgs e)
        {
            this.ValidateForControl(this.streetAddressTextBox);
        }

        private void cityTextBox_Validating(object sender, System.ComponentModel.CancelEventArgs e)
        {
            this.ValidateForControl(this.cityTextBox);
        }

        private void stateComboBox_Validating(object sender, System.ComponentModel.CancelEventArgs e)
        {
            this.ValidateForControl(this.stateComboBox);
        }

        private void zipCodeMaskedTextBox_Validating(object sender, System.ComponentModel.CancelEventArgs e)
        {
            this.ValidateForControl(this.zipCodeMaskedTextBox);
        }

        private void phoneMaskedTextBox_Validating(object sender, System.ComponentModel.CancelEventArgs e)
        {
            this.ValidateForControl(this.phoneMaskedTextBox);
        }

        private void emailAddressTextBox_Validating(object sender, System.ComponentModel.CancelEventArgs e)
        {
            this.ValidateForControl(this.emailAddressTextBox);
        }

        private void checkInAndCheckOutDateTimePickers_ValueChanged(object sender, EventArgs e)
        {
            this.ValidateForControl(this.checkInDateTimePicker);
            this.ValidateForControl(this.checkOutDateTimePicker);
        }

        private void rentalUnitsRadioButtons_Validating(object sender, System.ComponentModel.CancelEventArgs e)
        {
            this.ValidateForControl(this.rentalUnitsGroupBox);
        }
    }
}

File: cs205/jpetroulesCLASSES/jpetroulesCLASSESvb/MainForm.vb
' CONDO
' Jake Petroules
' 2010-10-11
' This program allows condominium rentals to be made at Mt. Snow

Option Strict On
Option Explicit On

Imports System.Globalization
Imports System.Net.Mail
Imports System.Text

Public Class MainForm
    ''' <summary>
    ''' Generates a receipt containing customer information, rental unit, rental period,
    ''' optional services and payment information and returns the result as a string.
    ''' </summary>
    Private Function GenerateReceipt() As String
        Dim builder As StringBuilder = New StringBuilder()

        ' Add the customer information
        builder.AppendLine(String.Format(CultureInfo.CurrentCulture, "{0} {1} {2}", Me.firstNameTextBox.Text, Me.middleNameTextBox.Text, Me.lastNameTextBox.Text))
        builder.AppendLine(Me.streetAddressTextBox.Text)
        builder.AppendLine(String.Format(CultureInfo.CurrentCulture, "{0}, {1} {2}", Me.cityTextBox.Text, Me.stateComboBox.Text, Me.zipCodeMaskedTextBox.MaskedTextProvider.ToString()))
        builder.AppendLine(Me.phoneMaskedTextBox.MaskedTextProvider.ToString())
        builder.AppendLine(Me.emailAddressTextBox.Text)
        builder.AppendLine()

        ' Add the rental unit information
        If Me.studioRadioButton.Checked Then
            builder.AppendLine(Me.studioRadioButton.Text)
        ElseIf Me.suite2RadioButton.Checked Then
            builder.AppendLine(Me.suite2RadioButton.Text)
        ElseIf Me.suite3RadioButton.Checked Then
            builder.AppendLine(Me.suite3RadioButton.Text)
        ElseIf Me.penthouseRadioButton.Checked Then
            builder.AppendLine(Me.penthouseRadioButton.Text)
        End If

        ' Add the rental period information (check-in/check-out times)
        Dim rentalPeriod As TimeSpan = Me.checkOutDateTimePicker.Value.Date - Me.checkInDateTimePicker.Value.Date
        builder.AppendLine(String.Format(CultureInfo.CurrentCulture, "Check in: {0}", Me.checkInDateTimePicker.Value.Date))
        builder.AppendLine(String.Format(CultureInfo.CurrentCulture, "Check out: {0}", Me.checkOutDateTimePicker.Value.Date))
        builder.AppendLine(String.Format(CultureInfo.CurrentCulture, "Rental period: {0} days", rentalPeriod.Days))
        builder.AppendLine()

        ' Add the optional services if they were selected
        If Me.maidServiceCheckBox.Checked OrElse Me.linenServiceCheckBox.Checked OrElse Me.limoServiceCheckBox.Checked Then
            builder.AppendLine("Optional services:")
        End If

        If Me.maidServiceCheckBox.Checked Then
            builder.AppendLine(Me.maidServiceCheckBox.Text)
        End If

        If Me.linenServiceCheckBox.Checked Then
            builder.AppendLine(Me.linenServiceCheckBox.Text)
        End If

        If Me.limoServiceCheckBox.Checked Then
            builder.AppendLine(Me.limoServiceCheckBox.Text)
        End If

        ' Add the customer's total bill
        builder.AppendLine(String.Format(CultureInfo.CurrentCulture, "Grand total: {0:C}", Me.CalculateTotal()))

        Return builder.ToString()
    End Function

    ''' <summary>
    ''' Calculates the customer's total bill and returns the result as a decimal
    ''' </summary>
    Private Function CalculateTotal() As Decimal
        Dim rentalPeriod As TimeSpan = Me.checkOutDateTimePicker.Value.Date - Me.checkInDateTimePicker.Value.Date

        ' Determine the weekly price of the selected rental unit
        Dim unit As RentalUnit
        If Me.studioRadioButton.Checked Then
            unit = RentalUnit.Studio
        ElseIf Me.suite2RadioButton.Checked Then
            unit = RentalUnit.TwoBedroomSuite
        ElseIf Me.suite3RadioButton.Checked Then
            unit = RentalUnit.ThreeBedroomSuite
        ElseIf Me.penthouseRadioButton.Checked Then
            unit = RentalUnit.Penthouse
        End If

        Dim services As OptionalServices = 0
        If Me.maidServiceCheckBox.Checked Then
            services = services Or OptionalServices.Maid
        End If
        
        If Me.linenServiceCheckBox.Checked Then
            services = services Or OptionalServices.Linen
        End If

        If Me.limoServiceCheckBox.Checked Then
            services = services Or OptionalServices.Limo
        End If

        Return RentalCalculator.CalculateTotal(rentalPeriod, unit, services)
    End Function

    ''' <summary>
    ''' Determines whether the user has entered complete and valid input in all fields.
    ''' </summary>
    Private Function IsInputValid() As Boolean
        Return _
            Me.ValidateForControl(Me.firstNameTextBox) And
            Me.ValidateForControl(Me.middleNameTextBox) And
            Me.ValidateForControl(Me.lastNameTextBox) And
            Me.ValidateForControl(Me.streetAddressTextBox) And
            Me.ValidateForControl(Me.cityTextBox) And
            Me.ValidateForControl(Me.stateComboBox) And
            Me.ValidateForControl(Me.zipCodeMaskedTextBox) And
            Me.ValidateForControl(Me.phoneMaskedTextBox) And
            Me.ValidateForControl(Me.emailAddressTextBox) And
            Me.ValidateForControl(Me.checkInDateTimePicker) And
            Me.ValidateForControl(Me.checkOutDateTimePicker) And
            Me.ValidateForControl(Me.rentalUnitsGroupBox)
    End Function

    ''' <summary>
    ''' Runs validation for the specified control and returns whether its corresponding input is valid.
    ''' </summary>
    ''' <param name="control">The control to validate.</param>
    Private Function ValidateForControl(ByVal control As Control) As Boolean
        Dim firstNameTextBox As TextBox = TryCast(control, TextBox)
        If firstNameTextBox Is Me.firstNameTextBox Then
            If String.IsNullOrWhiteSpace(Me.firstNameTextBox.Text) Then
                Me.firstNameErrorProvider.SetError(Me.firstNameTextBox, "Please enter a first name.")
                Return False
            Else
                Me.firstNameErrorProvider.Clear()
                Return True
            End If
        End If

        Dim middleNameTextBox As TextBox = TryCast(control, TextBox)
        If middleNameTextBox Is Me.middleNameTextBox Then
            If String.IsNullOrWhiteSpace(Me.middleNameTextBox.Text) Then
                Me.middleNameErrorProvider.SetError(Me.middleNameTextBox, "Please enter a middle name.")
                Return False
            Else
                Me.middleNameErrorProvider.Clear()
                Return True
            End If
        End If

        Dim lastNameTextBox As TextBox = TryCast(control, TextBox)
        If lastNameTextBox Is Me.lastNameTextBox Then
            If String.IsNullOrWhiteSpace(Me.lastNameTextBox.Text) Then
                Me.lastNameErrorProvider.SetError(Me.lastNameTextBox, "Please enter a last name.")
                Return False
            Else
                Me.lastNameErrorProvider.Clear()
                Return True
            End If
        End If

        Dim streetAddressTextBox As TextBox = TryCast(control, TextBox)
        If streetAddressTextBox Is Me.streetAddressTextBox Then
            If String.IsNullOrWhiteSpace(Me.streetAddressTextBox.Text) Then
                Me.streetAddressErrorProvider.SetError(Me.streetAddressTextBox, "Please enter a street address.")
                Return False
            Else
                Me.streetAddressErrorProvider.Clear()
                Return True
            End If
        End If

        Dim cityTextBox As TextBox = TryCast(control, TextBox)
        If cityTextBox Is Me.cityTextBox Then
            If String.IsNullOrWhiteSpace(Me.cityTextBox.Text) Then
                Me.cityErrorProvider.SetError(Me.cityTextBox, "Please enter a city.")
                Return False
            Else
                Me.cityErrorProvider.Clear()
                Return True
            End If
        End If

        Dim stateComboBox As ComboBox = TryCast(control, ComboBox)
        If stateComboBox Is Me.stateComboBox Then
            If Me.stateComboBox.SelectedIndex < 0 Then
                Me.stateErrorProvider.SetError(Me.stateComboBox, "Please select a state.")
                Return False
            Else
                Me.stateErrorProvider.Clear()
                Return True
            End If
        End If

        Dim zipCodeMaskedTextBox As MaskedTextBox = TryCast(control, MaskedTextBox)
        If zipCodeMaskedTextBox Is Me.zipCodeMaskedTextBox Then
            If Me.zipCodeMaskedTextBox.MaskCompleted Then
                Me.zipCodeErrorProvider.Clear()
                Return True
            Else
                Me.zipCodeErrorProvider.SetError(Me.zipCodeMaskedTextBox, "Please enter a valid ZIP code.")
                Return False
            End If
        End If

        Dim phoneMaskedTextBox As MaskedTextBox = TryCast(control, MaskedTextBox)
        If phoneMaskedTextBox Is Me.phoneMaskedTextBox Then
            If Me.phoneMaskedTextBox.MaskCompleted Then
                Me.phoneErrorProvider.Clear()
                Return True
            Else
                Me.phoneErrorProvider.SetError(Me.phoneMaskedTextBox, "Please enter a valid telephone number.")
                Return False
            End If
        End If

        Dim emailAddressTextBox As TextBox = TryCast(control, TextBox)
        If emailAddressTextBox Is Me.emailAddressTextBox Then
            Dim email As MailAddress
            Try
                email = New MailAddress(Me.emailAddressTextBox.Text)
                Me.emailAddressErrorProvider.Clear()
                Return True
            Catch argex As ArgumentException
                Me.emailAddressErrorProvider.SetError(Me.emailAddressTextBox, "Please enter an email address.")
                Return False
            Catch fex As FormatException
                Me.emailAddressErrorProvider.SetError(Me.emailAddressTextBox, "Please enter a valid email address.")
                Return False
            End Try
        End If

        Dim checkInDateTimePicker As DateTimePicker = TryCast(control, DateTimePicker)
        Dim checkOutDateTimePicker As DateTimePicker = TryCast(control, DateTimePicker)
        If checkInDateTimePicker Is Me.checkInDateTimePicker OrElse checkOutDateTimePicker Is Me.checkOutDateTimePicker Then
            Dim rentalPeriod As TimeSpan = Me.checkOutDateTimePicker.Value.Date - Me.checkInDateTimePicker.Value.Date
            If rentalPeriod.TotalDays > 0 Then
                Me.rentalTermLabel.Text = String.Format(CultureInfo.CurrentCulture, "You have selected a rental period of {0} days", rentalPeriod.Days)
                Me.rentalTermLabel.ForeColor = Label.DefaultForeColor
                Return True
            Else
                Me.rentalTermLabel.Text = "Check in time cannot be equal to or later than check out time"
                Me.rentalTermLabel.ForeColor = Color.Red
                Return False
            End If
        End If

        Dim rentalUnitsGroupBox As GroupBox = TryCast(control, GroupBox)
        If rentalUnitsGroupBox Is Me.rentalUnitsGroupBox Then
            Dim oneButtonChecked As Boolean = Me.studioRadioButton.Checked Xor Me.suite2RadioButton.Checked Xor Me.suite3RadioButton.Checked Xor Me.penthouseRadioButton.Checked
            If oneButtonChecked Then
                Me.rentalUnitsErrorProvider.Clear()
                Return True
            Else
                Me.rentalUnitsErrorProvider.SetError(Me.rentalUnitsGroupBox, "Please select a rental unit.")
                Return False
            End If
        End If

        ' An invalid control can just return false
        Return False
    End Function

    Private Sub generateReceiptButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles generateReceiptButton.Click, generateReceiptMenuItem.Click
        If Me.IsInputValid() Then
            Me.outputRichTextBox.Text = Me.GenerateReceipt()
        Else
            MessageBox.Show("Please correct the input errors before generating a receipt.", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
        End If
    End Sub

    Private Sub firstNameTextBox_Validating(ByVal sender As System.Object, ByVal e As System.ComponentModel.CancelEventArgs) Handles firstNameTextBox.Validating
        Me.ValidateForControl(Me.firstNameTextBox)
    End Sub

    Private Sub middleNameTextBox_Validating(ByVal sender As System.Object, ByVal e As System.ComponentModel.CancelEventArgs) Handles middleNameTextBox.Validating
        Me.ValidateForControl(Me.middleNameTextBox)
    End Sub

    Private Sub lastNameTextBox_Validating(ByVal sender As System.Object, ByVal e As System.ComponentModel.CancelEventArgs) Handles lastNameTextBox.Validating
        Me.ValidateForControl(Me.lastNameTextBox)
    End Sub

    Private Sub streetAddressTextBox_Validating(ByVal sender As System.Object, ByVal e As System.ComponentModel.CancelEventArgs) Handles streetAddressTextBox.Validating
        Me.ValidateForControl(Me.streetAddressTextBox)
    End Sub

    Private Sub cityTextBox_Validating(ByVal sender As System.Object, ByVal e As System.ComponentModel.CancelEventArgs) Handles cityTextBox.Validating
        Me.ValidateForControl(Me.cityTextBox)
    End Sub

    Private Sub stateComboBox_Validating(ByVal sender As System.Object, ByVal e As System.ComponentModel.CancelEventArgs) Handles stateComboBox.Validating
        Me.ValidateForControl(Me.stateComboBox)
    End Sub

    Private Sub zipCodeMaskedTextBox_Validating(ByVal sender As System.Object, ByVal e As System.ComponentModel.CancelEventArgs) Handles zipCodeMaskedTextBox.Validating
        Me.ValidateForControl(Me.zipCodeMaskedTextBox)
    End Sub

    Private Sub phoneMaskedTextBox_Validating(ByVal sender As System.Object, ByVal e As System.ComponentModel.CancelEventArgs) Handles phoneMaskedTextBox.Validating
        Me.ValidateForControl(Me.phoneMaskedTextBox)
    End Sub

    Private Sub emailAddressTextBox_Validating(ByVal sender As System.Object, ByVal e As System.ComponentModel.CancelEventArgs) Handles emailAddressTextBox.Validating
        Me.ValidateForControl(Me.emailAddressTextBox)
    End Sub

    Private Sub checkInAndCheckOutDateTimePickers_ValueChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles checkInDateTimePicker.ValueChanged, checkOutDateTimePicker.ValueChanged
        Me.ValidateForControl(Me.checkInDateTimePicker)
        Me.ValidateForControl(Me.checkOutDateTimePicker)
    End Sub

    Private Sub rentalUnitsRadioButtons_Validating(ByVal sender As System.Object, ByVal e As System.ComponentModel.CancelEventArgs) Handles studioRadioButton.Validating, suite2RadioButton.Validating, suite3RadioButton.Validating, penthouseRadioButton.Validating
        Me.ValidateForControl(Me.rentalUnitsGroupBox)
    End Sub

    Private Sub exitMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles exitMenuItem.Click
        Application.Exit()
    End Sub

    Private Sub aboutMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles aboutMenuItem.Click
        Dim dialog As AboutBox = New AboutBox
        dialog.ShowDialog()
    End Sub
End Class
